### PR TITLE
[FLINK-26985][runtime] Don't discard shared state of restored checkpoints

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

As described in the ticket, in `LEGACY` restore mode,
shared state of incremental checkpoints can be discarded
regardless of whether they were created by this job or not.

The bug was introduced in FLINK-24611. Before, reference count was maintained
for each entry;
"initial" checkpoints did not decrement this count, preventing their shared state from being discarded.

This PR makes `SharedStateRegistry` to:
1. remember the max checkpiont ID encountered during recovery
2. associate each state entry with a checkpoint ID that created it
3. only discard the entry if its `createdByCheckpointID` > highestRetainCheckpointID``

(1) is called from:
- `CheckpointCoordinator.restoreSavepoint` - to cover initial restore from a checkpoint
- `SharedStateFactory`, when building checkpoint store - to cover the failover case

Only `CheckpointCoordinator` does not seem sufficient, because a new checkpoint
can be created, from which the job can recover automatically, without calling `restoreSavepoint`.

(see `DefaultExecutionGraphFactory.createAndRestoreExecutionGraph`)

## Verifying this change

`ResumeCheckpointManuallyITCase` in `LEGACY` restore mode

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:270]: This field is not included in the `#toString()` method

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:187]: Why only `CLAIM` mode does not need to update the `highestRetainCheckpointID`? I think this deserve a description.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:57]: I think `highestRetainCheckpointID` might not be a good choice as we could still retain multi checkpoints even in `CLAIM` restore mode. However, current implementation would still keep the `highestRetainCheckpointID` as `-1` in `CLAIM` restore mode.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRecoveryFactory.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRecoveryFactory.java:41]:
```suggestion
     * @param restoreMode the restore mode with which the job is restoring.
```

**Review by Myasuka — COMMENTED** [review]
I think current solution looks a bit friable as it involes many changes to different APIs.
I left some comments, and please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:187]: In `CLAIM` mode, the state of the checkpoints that the job is recovering from can be deleted (must be deleted).
I'll add this comment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:57]: Could you explain which case do you mean and what would be the problem?

With `CLAIM` mode, the goal is to discard the state, including retained checkpoints. So this field `highestRetainCheckpointID` must be `-1`, which will allow to do this (condition in state entry will be met).

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/ResumeCheckpointManuallyITCase.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/ResumeCheckpointManuallyITCase.java:332]: You're right, I'll revert this change.

**Comment by rkhachatryan** [comment]
Thanks for the review @Myasuka ,

> I think current solution looks a bit friable as it involes many changes to different APIs.

I'm not sure what exactly do you mean. Indeed, there are changes to multiple interfaces, mostly because of the need to pass `RestoreMode` from `CheckpointRecoveryFactory` to `SharedStateRegistryFactory`.  However, I think this is unavoidable. Or do you see any alternative?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: Just as what I asked in the above, can we ensure this method only be called during restore from checkpoint/savepoint?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:57]: I feel that the naming of `retain` could make user mixed with the configuration `state.checkpoints.num-retained`. How about `reservedCheckpointIdOnRestore`, and we give clear description that `-1` means no checkpoint needs to be reserved on restore (actualy CLAIM mode).

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for the update.
I think we should analyze the lifecycle of `SharedStateRegistry#registerAllAfterRestored` to ensure it is call in the right place.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:40]: Yes, the shared state in that case won't be discarded. 
This mirrors the old behavior (pre-FLINK-24611) IIUC.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: I don't think it's possible.
If the above call is removed (from `SharedStateRegistryFactory`),
and only the call from `CheckpointCoordinator` is left,

then in case of JM failover **before** taking any new checkpoints, the old no-claim-state will be discarded.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/RestoreMode.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/RestoreMode.java]: Thanks for noticing that!
I've changed it to `NO_CLAIM` in 07d732d04d44081880e248a553bcc6f653d6e53e.

**Review by akalash — COMMENTED** [review]
@rkhachatryan, I like the idea but I have a couple of questions/suggestions about implementation which I want to discuss. So please, take a look at my comments

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:40]: My intention was to propagate the restore mode from the initial job settings, so it should NOT be DEFAULT
(and most of the changes are caused by that propagation).

I see that in the current PR, `RestoreMode.DEFAULT` is only used by `CheckpointResourcesCleanupRunner`, all the other usages eventually use `jobGraph.getSavepointRestoreSettings().getRestoreMode()`. Or am I missing something?

Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:40]: Oh, I see that `SchedulerUtils` uses `RestoreMode` from the settings. Then indeed my comment is not relevant and we don't use only DEFAULT here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:185]: Because a single completed checkpoint should not be aware of the `highestRetainCheckpointID`; `SharedStateRegistry` would have to have a separate method to update `highestRetainCheckpointID`. 
IMO, that would make the code less readable and more error-prone.

I'm also not sure whether `SharedStateRegistry` should be aware of the restore mode or not;
But given that there are no obvious reasons against it, and taking the above into account, I'd stick with the current approach.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:40]: Thank you for double-checking

Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:40]: What I mean is that if we restore from chk-10 with `NO_CLAIM` mode, and then the JM crashed on chk-20, will the chk-20 also be treated as the not-claimed checkpoint, leading to the chk-20 never be discarded anymore. If so, does this make orphan checkpoints left?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: What will happen if JM crashed after newer checkpoints taken? Does it mean we have to restore the newly created checkpoints and update the `highestNotClaimedCheckpointID` to the newer created checkpoint, which leads to the new created checkpoints not discarded anymore?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:40]: Yes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:185]: Perhaps, I wasn't fully correct in my comment. But I mean that `highestRetainCheckpointID` as field of `SharedStateRegistry` is ok. But instead of `SharedStateRegistry#registerAllAfterRestored(CompletedCheckpoint checkpoint, RestoreMode mode)` we can use following signature `registerAllAfterRestored(CompletedCheckpoint checkpoint, long retainCheckpointID)`. And inside  `CompletedCheckpoint#registerSharedStatesAfterRestored` we can call `registerAllAfterRestored' with the current checkpoint id if the restore mode is not claim and with -1 otherwise.

But I agree that it is a minor change and it is up to you. I just don't really sure that `SharedStateRegistryImpl`should know something about RestoreMode.

**Review by akalash — APPROVED** [review]
LGTM

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: Yes.

Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: You're proposing to allow discarding the state if the latest checkpoint path is different from that of `SavepointRestoreSettings.getRestorePath`, right?

I think it would again invalidate the initial checkpoint, because the latest checkpoint might still be using the initial shared state.

Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: I think current dilemma is caused by that we might not make the refactory of equivalence.

Previously, no matter whether JM crashed or just restore from checkpoint/savepoint. All checkpoints would be registered to incrase the reference counting (via `CompletedCheckpoint#registerSharedStatesAfterRestored`), and the checkpoint property of `discardOnSubume` could help whether to discard this registered checkpoint to decrease the reference on subsume. 
However, this changed after FLINK-24611. Previously, the unregister phase for SharedStateReigstry does not depend on the checkpoint id relationship, we just unregister the checkpoints independently and not consider which checkpoint id is larger. However, after FLINK-24611, this changed and I think we could persist the restored id in some place to mitigate such regression.

Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: > state left-over in case of JM failover is not a regression 

I think this is only ture for JM crash during on-going checkpoint discard. In other cases, this is a regression, although we had to say that the probability of JM crash is really small.

Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: > Previously, no matter whether JM crashed or just restore from checkpoint/savepoint. All checkpoints would be registered to incrase the reference counting (via CompletedCheckpoint#registerSharedStatesAfterRestored)

If the initial checkpoints were replaced (subsumed) by the new ones; then `discardOnSubume` property would be lost, wouldn't it?
Then the new JM would load the new checkpoints with `discardOnSubume=true` and after some time discard the initial shared state.

Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: I think your analysis is right before Flink-1.15, and Flink actually did not give exact defination on the restore mode before Flink-1.15.

How about persisting the restored checkpoint id in the checkpoint store and update the restored checkpoint id on each time checkpoint coordinator to call restoreSavepoint?

Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: In general, we should persist the restore mode and restored checkpoint id in the checkpoint store. With these information, we can make the restore and discard phase more safe.

**Comment by curcur** [comment]
I am mostly fine with the changes. Besides the comments left above, my main concern is why introducing "RestoreMode DEFAULT = NO_CLAIM", let me explain my concerns in the following:

1. I am fine with the changes along the path "CheckpointCoordinator#restoreSavepoint", including all the interface changes, e.t.c.
2. I am not fine with `RestoreMode DEFAULT = NO_CLAIM` and passed in `EmbeddedCompletedCheckpointStore`, `StandaloneCompletedCheckpointStore`, `CheckpointResourcesCleanupRunner`, `EmbeddedHaServicesWithLeadershipControl`. There, it should follow the normal failover paths assuming all checkpoints belong to the same job, and non-referenced are subsumed as normal. I do not think NO_CLAIM follows this way.
So I would suggest we introduce a fourth internal value to indicate the normal JM failover behavior
3. @Myasuka mentioned a good point that item 2 itself is not enough: it is possible that before the first checkpoint after recovery is able to be made, JM crashes again, so item 2 may lead to losing information about whether the previous checkpoint is from a previous job or previous run.

We can also discuss offline.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:185]: Thanks for clarifying. Indeed, a separate method would not be needed with this approach. But now the logic inside `CompletedCheckpoint` will be tricky IMO: `registry.registerAll(..., mode == CLAIM ? -1 : this.id`. So I think I'll keep the logic in place in `SharedStateRegistry` for now.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: I agree that a restored checkpoint ID should be persisted durably somewhere. 
And it can also be subsequent checkpoints themselves. I think we need some more analysis to chose the right option and implement correctly. 

And because this is not strictly a regression, I strongly prefer to fix it separately and also not block the release.
Another aspect is that we should address this in 1.14 too (for `LEGACY` mode).

WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/RestoreMode.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/RestoreMode.java:57]: Good point, I think I'll change the `defaultValue` of `SavepointConfigOptions.RESTORE_MODE` to `RestoreMode.DEFAULT`:
```
public static final ConfigOption<RestoreMode> RESTORE_MODE =
        key("execution.savepoint-restore-mode")
                .enumType(RestoreMode.class)
                .defaultValue(RestoreMode.DEFAULT)
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:192]: Although the initial checkpoint is fully copied in `NO_CLAIM` mode, only the **next** checkpoint can be "claimed". 
`SharedStateRegistry ` and `CompletedCheckpointStore` still refer to the initial checkpoint until it is subsumed.
So we must prevent it's artifacts from being deleted.

Comment (curcur) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:192]: "initial checkpoint is fully copied" and owned by the current job, why it can not follow the normal subsumed way of other checkpoints?

So how the first full checkpoint is different from the second (next) checkpoint?

Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java:192]: > "initial checkpoint is fully copied" and owned by the current job, why it can not follow the normal subsumed way of other checkpoints?

No, initial checkpoint is NOT owned by the current job. When performing a **new** checkpoint after recovery, initial state is copied (if needed) by state backends and that copied state is included into the **new** checkpoint.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:88]: I think the issue deserves its own ticket regardless of when we decide to address it, so I've created FLINK-27114 to track it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java:216]: Yes, I think this is the same question as above.

**Comment by curcur** [comment]
> Thanks for the review @curcur, regarding
> 
> > I am not fine with RestoreMode DEFAULT = NO_CLAIM and passed in EmbeddedCompletedCheckpointStore, StandaloneCompletedCheckpointStore, CheckpointResourcesCleanupRunner, EmbeddedHaServicesWithLeadershipControl.
> 
> I've updated `EmbeddedHaServicesWithLeadershipControl` to use the `RestoreMode` from job settings. Other cases are test-only or specific to cleanup, where the restore mode is not applicable.
> 
> > assuming all checkpoints belong to the same job, and non-referenced are subsumed as normal
> 
> I think this depends on [#19331 (comment)](https://github.com/apache/flink/pull/19331#discussion_r844665419)

Who will clean-up the checkpoints if JM failover, and with no-claim mode?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/EmbeddedCompletedCheckpointStore.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/EmbeddedCompletedCheckpointStore.java:66]: If we changed the default restore mode to `CLAIM`, can we also keep this here?
From my understanding, we use `NO_CLAIM` mode for the safety to not discard checkpoint.

And if we just want to make the safety, I think this deserves a description and maybe we need to refer to FLINK-27114.

**Review by Myasuka — COMMENTED** [review]
I only left one additional comment, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/EmbeddedCompletedCheckpointStore.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/EmbeddedCompletedCheckpointStore.java]: I've added a comment here and in `StandaloneCompletedCheckpointStore` after an offline discussion

**Comment by Myasuka** [comment]
I think current PR is in a good shape now. @XComp could you please double confirm the change within [CheckpointResourcesCleanupRunner.java](https://github.com/apache/flink/pull/19331/files#diff-1e3a73b501a11e2af7332a39c7e7f63c57a632a601703ae5574be1f9e9de244dR122-R126) ?

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/CheckpointResourcesCleanupRunner.java
Comment (XComp) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/CheckpointResourcesCleanupRunner.java:126]: Could we move the `RestoreMode` into the private method `createCompletedCheckpointStore`? At least, I don't see a reason to introduce a parameter here...

**Review by XComp — APPROVED** [review]
Just a minor cosmetic thing from my end. But other than that, using `CLAIM` as the `RestoreMode` seem to be the right thing to do because we actually want to clean up the all the checkpoints in this runner.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/CheckpointResourcesCleanupRunner.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/CheckpointResourcesCleanupRunner.java]: I agree, inlined the argument.

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java:313]: The comment is out of date once you introduce the parameter of `numCheckpoints`.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/ResumeCheckpointManuallyITCase.java
Comment (Myasuka) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/ResumeCheckpointManuallyITCase.java:348]: Add a comment to describe we need at least 2 checkpoints here.

**Review by Myasuka — APPROVED** [review]
LGTM except two minor suggestions.

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java]: I've update the comment.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/ResumeCheckpointManuallyITCase.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/ResumeCheckpointManuallyITCase.java:349]: Added the comment.

**Comment by rkhachatryan** [comment]
Thanks for the thorough review @Myasuka, @curcur, @akalash  and @XComp.
I've addressed the latest remarks and will merge the PR once the build completes.

**Comment by rkhachatryan** [comment]
Build failures unrelated:
- PR: FLINK-27148
- private branch: FLINK-27148, FLINK-27031
However, `finegrained_resource_management` phase failed in both builds. I"ve restarted the failed stages in private build.

**Comment by rkhachatryan** [comment]
@flinkbot run azure

**Comment by rkhachatryan** [comment]
Build succeeded, merging.

**Pull request merged**
